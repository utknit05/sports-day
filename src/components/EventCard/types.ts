import { IEvent } from "../../context/EventsList/types";

type TEventActionHandler = (event: IEvent) => void

export type TEventCardProps =
    { event: IEvent } &
    ({
        type: 'selected';
        onRemove: TEventActionHandler;
    } | {
        type: 'unselected';
        onSelect: TEventActionHandler;
    })

export enum EEventsCategory {
    Swimming = 'Swimming',
    Athletics = 'Athletics',
    Boxing = 'Boxing',
}

export const categoryToGradientMap: {
    [key in EEventsCategory]: string
} = {
    [EEventsCategory.Swimming]: 'linear-gradient(to left, rgb(27, 214, 255), rgb(59, 157, 255))',
    [EEventsCategory.Athletics]: 'linear-gradient(to right, rgb(236, 97, 23, 90), rgb(0, 108, 255, 90))',
    [EEventsCategory.Boxing]: 'linear-gradient(to right, rgb(236, 58, 23), rgb(255, 255, 255))',
}
