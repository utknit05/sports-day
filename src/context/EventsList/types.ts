import { IListResponse } from "../../utils/api";
import { fetchEvents, fetchEventsCompleted, resetEvents } from "./store";

export const GET_EVENTS_ENDPOINT = 'https://run.mocky.io/v3/2744c231-8991-4ae8-bc45-1f645437585a';

export interface IEvent {
  id: number;
  event_name: string;
  event_category: string;
  start_time: string;
  end_time: string;
  disabled?: boolean;
}

export type TEventsState = IListResponse<IEvent>
export type TEventsActions = ReturnType<typeof fetchEvents | typeof fetchEventsCompleted | typeof resetEvents>
export interface IEventsContext {
  selectedEvents: IEvent[];
  unselectedEvents: IEvent[];
  selectEvent: (event: IEvent) => void
  removeEvent: (event: IEvent) => void
}
