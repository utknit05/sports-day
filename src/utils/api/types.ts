export enum ERequestState {
    INITIAL,
    REQUESTING,
    SUCCESS,
    FAILED,
}

export interface IListResponse<T>  {
    list: T[];
    requestState: ERequestState;
    error?: {
        code: string;
        message: string;
    };
}
