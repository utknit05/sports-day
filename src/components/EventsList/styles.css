.eventsListContainer {
    margin-bottom: 24px;
    width: 100%;
}

.eventsListLayout {
    display: grid;
    grid-auto-rows: auto;
    grid-template-columns: 280px 280px 280px;
    gap: 10px;
    padding-left: 16px;
}

@media only screen and (max-width: 1000px) {
    .eventsListLayout {
        grid-template-columns: 280px 280px;
    }   
}

@media only screen and (max-width: 600px) {
    .eventsListLayout {
        grid-template-columns: 280px;
    }   
}

.filtersContainer {
    display: flex;
    padding: 8px 14px;
}
