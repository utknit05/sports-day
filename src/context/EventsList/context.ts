import { createContext, useContext } from 'react'
import { IEventsContext } from './types'

export const EventsStateContext = createContext<IEventsContext | null>(null)

export function useEventsState() {
    const context = useContext(EventsStateContext)
    if (!context) {
        throw new Error('useEventsState must be used within EventsStateContext')
    }
    return context
}
