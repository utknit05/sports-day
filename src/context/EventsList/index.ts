export { default as EventsProvider } from './Provider';
export { useEventsState } from './context';
export { fetchEvents, fetchEventsCompleted, resetEvents } from './store';
