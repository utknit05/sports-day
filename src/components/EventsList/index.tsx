import React, { useState } from 'react';
import { useEventsState } from '../../context/EventsList';
import EventCard from '../EventCard';
import './styles.css';
import FilterPill from '../FilterPill';
import Header from '../Header';

interface IEventsListProps {
    type: 'selected' | 'unselected';
}

const EventsList: React.FC<IEventsListProps> = ({ type }) => {
    const { selectedEvents, unselectedEvents, removeEvent, selectEvent } = useEventsState();
    const [selectedFilter, setSelectedFilter] = useState<string>();

    const eventsFilters = type === 'selected' ? new Set(selectedEvents.map(({ event_category }) => event_category)) : new Set(unselectedEvents.map(({ event_category }) => event_category));

    return (
        <div className="eventsListContainer">
            <Header title={type === 'selected' ? "Selected Events" : "All Events"} />
            {
                type === 'selected' ? (
                    <>
                        <div className="filtersContainer">
                            <FilterPill title="All" onClick={setSelectedFilter} isSelected={!selectedFilter} />
                            {Array.from(eventsFilters).map(category => <FilterPill key={category} id={category} title={category} onClick={setSelectedFilter} isSelected={category === selectedFilter} />)}
                        </div>
                        <div className='eventsListLayout'>
                            {selectedEvents.filter(e => selectedFilter ? e.event_category === selectedFilter : true).map(e => {
                                return <EventCard key={e.id} event={e} type='selected' onRemove={removeEvent} />
                            })}
                        </div>
                    </>
                ) : (
                    <>
                        <div className="filtersContainer">
                            <FilterPill title="All" onClick={setSelectedFilter} isSelected={!selectedFilter} />
                            {Array.from(eventsFilters).map(category => <FilterPill key={category} id={category} title={category} onClick={setSelectedFilter} isSelected={category === selectedFilter}  />)}
                        </div>
                        <div className='eventsListLayout'>
                            {unselectedEvents.filter(e => selectedFilter ? e.event_category === selectedFilter : true).map(e => {
                                return <EventCard key={e.id} event={e} type='unselected' onSelect={selectEvent} />
                            })}
                        </div>
                    </>
                )
            }
        </div>
    )
}

export default EventsList;
