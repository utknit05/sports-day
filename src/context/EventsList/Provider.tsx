import React, { PropsWithChildren, useEffect, useReducer, useRef, useState } from 'react'
import { EventsStateContext } from './context';
import { fetchEvents, fetchEventsCompleted, initialState, reducer } from './store';
import { ERequestState, GENERIC_ERROR, callApi } from '../../utils/api';
import { GET_EVENTS_ENDPOINT, IEvent } from './types';
import { getConflictingEventsMap } from '../../utils/common';
import { getDisabledEventsInUnselectedList } from './utils';

const Provider: React.FC<PropsWithChildren> = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    const [selectedEvents, setSelectedEvents] = useState<IEvent[]>([]);
    const [unselectedEvents, setUnselectedEvents] = useState<IEvent[]>([]);
    const conflictingEventsRef = useRef<{ [key: string]: number[] }>();

    useEffect(() => {
        dispatch(fetchEvents());
        callApi.get(GET_EVENTS_ENDPOINT).then(res => {
            if (Array.isArray(res)) {
                dispatch(fetchEventsCompleted({
                    list: res,
                }));
                conflictingEventsRef.current = getConflictingEventsMap(res);
            } else {
                dispatch(fetchEventsCompleted({
                    error: GENERIC_ERROR
                }));
            }
        }).catch(e => {
            dispatch(fetchEventsCompleted({
                error: GENERIC_ERROR
            }));
            console.log(e);
        })
    }, [])

    useEffect(() => {
        if (state.requestState === ERequestState.SUCCESS) {
            const unselected = state.list.filter((event) => !selectedEvents.find(({ id }) => id === event.id));
            setUnselectedEvents(unselected);
        }
    }, [state.requestState])

    const selectEvent = (event: IEvent) => {
        if (selectedEvents.length === 3) {
            alert(`Maximum events that you can opt for has been exceeded. Remove atleast one of already selected event to add ${event.event_name}`)
            return;
        }
        const newSelectedEvents = [...selectedEvents, event];
        const newUnselectedEvents = getDisabledEventsInUnselectedList(
            newSelectedEvents,
            unselectedEvents.filter(e => e.id !== event.id),
            conflictingEventsRef.current,
        );
        setSelectedEvents(newSelectedEvents);
        setUnselectedEvents(newUnselectedEvents);
    }

    const removeEvent = (event: IEvent) => {
        const newSelectedEvents = selectedEvents.filter(e => e.id !== event.id);
        const newUnselectedEvents = getDisabledEventsInUnselectedList(
            newSelectedEvents,
            [...unselectedEvents, event],
            conflictingEventsRef.current,
        );
        setSelectedEvents(newSelectedEvents);
        setUnselectedEvents(newUnselectedEvents);
    }

    return (
        <EventsStateContext.Provider value={{ selectedEvents, unselectedEvents, selectEvent, removeEvent }}>
            {children}
        </EventsStateContext.Provider>
    )
}

Provider.displayName = 'EventsContextProvider';

export default Provider;
