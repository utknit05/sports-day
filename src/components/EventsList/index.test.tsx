import { render } from "@testing-library/react"
import { EventsStateContext } from "../../context/EventsList/context"
import EventsList from "."
import { IEvent } from "../../context/EventsList/types";

describe('EventsList', () => {
    const mockedOnRemove = jest.fn();
    const mockedOnSelect = jest.fn();
    const events: IEvent[] = [
        {
            id: 3,
            event_name: "Freestyle 400M",
            event_category: "Swimming",
            start_time: "2022-12-17 15:00:00",
            end_time: "2022-12-17 16:00:00"
        }
    ]

    it('renders selected EventsList', () => {
        const { container } = render((
            <EventsStateContext.Provider value={{ selectedEvents: events, unselectedEvents: [], removeEvent: mockedOnRemove, selectEvent: mockedOnSelect }}>
                <EventsList type="selected" />
            </EventsStateContext.Provider>
        ))

        expect(container).toMatchSnapshot();
    })

    it('renders unselected EventsList', () => {
        const { container } = render((
            <EventsStateContext.Provider value={{ selectedEvents: [], unselectedEvents: events, removeEvent: mockedOnRemove, selectEvent: mockedOnSelect }}>
                <EventsList type="unselected" />
            </EventsStateContext.Provider>
        ))

        expect(container).toMatchSnapshot();
    })
})
