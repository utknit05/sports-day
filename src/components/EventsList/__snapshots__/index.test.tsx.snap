// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsList renders selected EventsList 1`] = `
<div>
  <div
    class="eventsListContainer"
  >
    <div
      class="headerContainer"
    >
      Selected Events
    </div>
    <div
      class="filtersContainer"
    >
      <div
        class="filterPillContainer selectedFilterPill"
      >
        All
      </div>
      <div
        class="filterPillContainer "
      >
        Swimming
      </div>
    </div>
    <div
      class="eventsListLayout"
    >
      <div
        class="eventCardContainer"
      >
        <div
          class="eventDetailsContainer"
        >
          <h1>
            S
          </h1>
          <div
            class="eventDescription"
          >
            <p
              class="eventName"
            >
              Freestyle 400M
            </p>
            <p
              class="eventSubText"
            >
              (
              Swimming
              )
            </p>
            <p
              class="eventSubText"
            >
              3 PM - 4 PM
            </p>
          </div>
        </div>
        <div
          class="eventButtonWrapper"
        >
          <button
            class="eventActionButton removeButton"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EventsList renders unselected EventsList 1`] = `
<div>
  <div
    class="eventsListContainer"
  >
    <div
      class="headerContainer"
    >
      All Events
    </div>
    <div
      class="filtersContainer"
    >
      <div
        class="filterPillContainer selectedFilterPill"
      >
        All
      </div>
      <div
        class="filterPillContainer "
      >
        Swimming
      </div>
    </div>
    <div
      class="eventsListLayout"
    >
      <div
        class="eventCardContainer"
      >
        <div
          class="eventDetailsContainer"
        >
          <h1>
            S
          </h1>
          <div
            class="eventDescription"
          >
            <p
              class="eventName"
            >
              Freestyle 400M
            </p>
            <p
              class="eventSubText"
            >
              (
              Swimming
              )
            </p>
            <p
              class="eventSubText"
            >
              3 PM - 4 PM
            </p>
          </div>
        </div>
        <div
          class="eventButtonWrapper"
        >
          <button
            class="eventActionButton selectButton"
          >
            Select
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
